proposals/0047-last-hard-fork-syscall.md

```diff
@@ -0,0 +1,94 @@
+slot. So hard fork is a tool to intentionally fork off the nodes not
+participating in the restart effort.
+
+Program developers may find it useful to know that a hard fork has recently
+occurred. This information can help them prevent arbitrage and liquidation
+caused by outdated oracle price account states. However, the cluster's restart
+process takes several hours; in that time, the world can change, causing asset
+prices to fluctuate. As the cluster updates the state of all accounts, malicious
+actors could take advantage of the delay by conducting trades using outdated
+prices, anticipating that they will soon catch up to the actual prices of
+real-world assets. Knowing that hard fork has been done recently, programs can
+manage these cases more appropriately.
+
+## Alternatives Considered
+
+No alternate considerations; we need to have the value of last hard fork slot
+while executing the program to correctly treat this case. We cannot have an
+account because then it should be updated just after the restart is successful,
+which will add complexity. The best way is to create a new syscall to get this
+information during execution of transaction, which will help us get last
+hardfork slot without any interface changes for the instruction.
+
+## Detailed Design
+
+### Creation of a new syscall
+
+The implementation of this syscall is pretty straitforward. In solana client all
+the hardforks for a cluster are stored in the bank structure. The last hard fork
+slot can be retrieved and then stored in invoke context, so that the executing
+program can access it.
+
+For other clients, we have to get the last hard fork slot information and make
+it accessible to the runtime of the program. If there is no hard fork done yet
+on the cluster we consider that the first hard for is at Slot `0`.
+
+### Overview of changes for solana client
+
+The hardfork data is available in the `Bank`. The structure `HardForks` contains
+a vector with all the previous hard forks. The vector is also sorted when we
+register a slot. So the last element of the vector is usually the last slot at
+which hard fork was done. We can get the last hard fork slot from the bank and
+pass it to invoke context structure. We can register new syscall in the file
+`definitions.rs` where other syscalls are defined.
+
+```rust
+define_syscall!(fn sol_get_last_hardfork_slot() -> Slot);
+```
+
+We can then return the data of last hardfork slot passed to the invoke context
+in this functions implementation.
+
+## Impact
+
+Programs will start using this new syscall to correctly address the security
+concerns during network restart. This will increase the reliability of solana
+cluster as whole and make programs developers more confident to handle edge
+cases during such extreme events.
+
+As the method is syscall the developers do not need to pass any new
+accounts or sysvars to the instruction to use this feature.
```

Comment:
For frontends interacting with on-chain programs, they may also need to know the restart slot. For example, if the Mango program halts all settling 1000 slots after a hard fork, the frontend may like to show a message "no settling currently allowed, please wait X slots".

For frontends to fetch the information, we can expose the hard fork through an additional RPC call or sysvar account that exposes the entire `HardForks` structure. I don't feel too strongly either way, but I would lean towards the sysvar account since it gives more info and reduces RPC bloat.

On the other hand, given @t-nelson 's comment about generalizing this for future restart procedures, a new RPC call may be better. However, I don't know if we have enough information about the future restart process to know if the concept of a "restart slot" is even applicable.

Any thoughts here?

Reply:
I feel like at this point we don't need an explicit RPC method or add a sysvar just for cleaner frontend handling of this condition. Outages aren't common so as long as the program protects itself a standard transaction simulation by the frontend will produce an (probably generic) error for the user in this exceptional case.

Reply:
they can already update the ui by querying whatever account is storing the corresponding "proceed from restart" signal/slot for their program. trying to do this with an explicit rpc method means they'd have to somehow track state in the client, which will be error-prone at best. the way i see this function being used is something like

```
    match instruction {
        PostOrder(order) => market.do_post_order(order),
        RemoveOrder(order) => market.do_remove_order(order),
        Match(order1, order2) => {
            if market.open_since_slot < sol_last_restart_slot() {
                return Err(MarketIsPostOnly);
            }
            market.do_match(order1, order2)
        },
        OpenMarket => market.do_open(clock.slot),
    }
```